runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,427 @@
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    List<CorfuStoreEntry<K, V, M>> getByIndex(@Nonnull Table<K, V, M> table,
+                                              @Nonnull final String indexName,
+                                              @Nonnull final I indexKey) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numGetByIndex.incrementAndGet();
+        return table.getByIndex(indexName, indexKey);
+    }
+
+    /**
+     * Gets the count of records in the table at a particular timestamp.
+     *
+     * @param table - the table whose count is requested.
+     * @return Count of records.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    int count(@Nonnull final Table<K, V, M> table) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numCounts.incrementAndGet();
+        return table.count();
+    }
+
+    /**
+     * Gets all the keys of a table.
+     *
+     * @param table - the table whose keys are requested.
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numKeySets.incrementAndGet();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param table Table< K, V, M > object on which the scan must be done.
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        txBegin(table); // Apply pending writes (if any) & begin corfu txn if thread hasn't begun one
+        table.numScans.incrementAndGet();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the actual corfu transaction if started by a read operation,
+     * apply all pending operations after starting the corfu transaction.
+     * @param tableBeingRead - only read transactions set this value
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void txBegin(Table<K, V, M> tableBeingRead) {
+        if (tableBeingRead != null) { // called from one of the read api
+            if (!operations.isEmpty()) { // We have buffered some writes before..
+                operations.forEach(Runnable::run); // apply them for dirty reads to see
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+            }
```

Comment:
The logic here seems not correct. If this is the first time to call txBegin with tableBeingRead != null, since the transaction is not started yet, there is no TransactionalContext, but you still run the operations on line 344, which means they will be non-transactional writes. Did I miss anything?

Reply:
good catch - I was trying to refactor to avoid repetition and ended up missing this case.
The implementation is still filled with if - else logic, so please feel free to suggest a better alternative.